### PR TITLE
Wip rgw blake2

### PR DESCRIPTION
**Pull request by mattbenjamin**
## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -3910,10 +3931,17 @@ void RGWPutObj::execute()
       break;
     }
 
+    /* XXX likely this forces a rebuild of data--should implement a
+     * segmented Digest::Update */
```
Comment (cbodley) [inline src/rgw/rgw_op.cc]: good catch! i see there's a new templated `ceph::crypto::digest<Digest>(const ceph::buffer::list& bl)` helper function in ceph_crypto.h that calls Update() piecewise. maybe we could split that logic (without the init/finalize) into another helper like this?
```c++
template <typename DigestT>
void digest_update(DigestT& gen, const ceph::buffer::list& bl)
{
  for (auto& p : bl.buffers()) {
    gen.Update((const unsigned char *)p.c_str(), p.length());
  }
}
```

File: src/common/ceph_crypto.h [hunk 2]
```diff
@@ -45,8 +46,41 @@ namespace ceph {
...
+      }
+
+      void Final(unsigned char* digest) {
+	blake2bp_final(&s, digest, digest_size);
+      }
+    };
+
     namespace ssl {
-      class OpenSSLDigest {
+      class OpenSSLDigest : public Digest {
```
Comment (cbodley) [inline src/common/ceph_crypto.h]: should `class HMAC` below also inherit this?

File: src/common/ceph_crypto.h [hunk 3]
```diff
@@ -45,8 +46,41 @@ namespace ceph {
...
+      blake2bp_state s;
+    public:
+      static constexpr uint16_t digest_size =
+	BLAKE2B_OUTBYTES /* 64 */;
+
+      Blake2B() {
+	Restart();
+      }
+
+      void Restart() {
```
Comment (cbodley) [inline src/common/ceph_crypto.h]: can you please add `override` to help document that these are all virtual?

File: src/rgw/rgw_common.h [hunk 4]
```diff
@@ -1396,6 +1397,18 @@ inline ostream& operator<<(ostream& out, const RGWBucketIndexType &index_type)
   }
 }
 
+namespace rgw { namespace cksum {
```
Comment (cbodley) [inline src/rgw/rgw_common.h]: c++17 lets you open nested namespaces, ie `namespace rgw::cksum {`

File: src/rgw/rgw_common.h [hunk 5]
```diff
@@ -1444,7 +1459,7 @@ struct RGWBucketInfo {
   RGWObjectLock obj_lock;
 
   void encode(bufferlist& bl) const {
-     ENCODE_START(20, 4, bl);
+     ENCODE_START(21, 4, bl);
```
Comment (cbodley) [inline src/rgw/rgw_common.h]: there's a corresponding `DECODE_START_LEGACY_COMPAT_LEN_32(20, 4, 4, bl);` on decode() that should be updated too

File: src/rgw/rgw_op.h [hunk 6]
```diff
@@ -45,6 +45,7 @@
 #include "rgw_lc.h"
 #include "rgw_torrent.h"
 #include "rgw_tag.h"
+#include "rgw_cksum.h"
```
Comment (cbodley) [inline src/rgw/rgw_op.h]: the only other addition to this file is `string cksum;` so i don't think this include is needed

File: src/common/options.cc [hunk 7]
```diff
@@ -7171,6 +7171,13 @@ std::vector<Option> get_rgw_options() {
     .add_see_also("rgw_dmclock_metadata_res")
     .add_see_also("rgw_dmclock_metadata_wgt"),
 
+    Option("rgw_cksum_type", Option::TYPE_STR, Option::LEVEL_ADVANCED)
+    .set_default("")
+    .set_description("Secondary cryptographic digest algorithm")
+    .set_long_description(
+        "One of SHA256, SHA512, or Blake2B.  If set, the corresponding digest "
+        "is used to sign uploaded object data, and returned in custom header "
+        "x-rgw-checksum in GET and HEAD requests."),
```
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/common/options.cc]: cool, will add it

File: src/rgw/rgw_op.cc (hunk 1 above)
Comment (mattbenjamin) [reply src/rgw/rgw_op.cc]: cool, didn't know it existed, will update to use it

File: src/rgw/rgw_common.h (hunk 4 above)
Comment (mattbenjamin) [reply src/rgw/rgw_common.h]: I actually find opening each namespace easier to read

File: src/rgw/rgw_common.h (hunk 5 above)
Comment (mattbenjamin) [reply src/rgw/rgw_common.h]: oops, that would do it

File: src/rgw/rgw_op.h (hunk 6 above)
Comment (mattbenjamin) [reply src/rgw/rgw_op.h]: true, it's all in the .cc now

File: src/common/ceph_crypto.h [hunk 8]
```diff
@@ -239,14 +239,21 @@ namespace ceph {
...
 namespace ceph::crypto {
+
+template<class Digest>
+inline void update(Digest& d, const ceph::buffer::list& bl)
+{
+  for (auto& p : bl.buffers()) {
+    d.Update((const unsigned char *)p.c_str(), p.length());
+  }
+}
+
```
Comment (adamemerson) [inline src/common/ceph_crypto.h]: I approve of this function.

**Comment by cbodley** [comment]
needs a rebase

**Comment by mattbenjamin** [comment]
unstale

**Comment by mattbenjamin** [comment]
unstale

**Comment by mattbenjamin** [comment]
unstale

**Comment by mattbenjamin** [comment]
unstale

**Comment by mattbenjamin** [comment]
let this not be stale

**Comment by mattbenjamin** [comment]
oh noes!

**Pull request closed without merging**